### PR TITLE
AX: Implement loop_counter and support UCodes without LPF

### PR DESCRIPTION
**Pull request by MerryMage**
~~Uses constexpr if and hence requires C++17. This approach is possible without constexpr if but would be much uglier.~~

~~I decided to replace the preprocessor `AX_GC` and `AX_WII` defines with a configuration template parameter. This allowed for further variants in AX functionality to be supported via constexpr if.~~

~~While one can now semi-trivially remove the need for the global variables in `AXVoice.h` after this change, I've decided to leave that to another PR.~~

(P.S.: [delroth did all the hard work](https://bugs.dolphin-emu.org/issues/7228), this just implements the fix.)

Edit: Now uses copies for layout differences.

### CODE REVIEW

**Comment by MerryMage** [comment]
* Requires at least cmake 3.8-rc4 for C++17 support.
* Ugly formatting due to clang-format's lack of support for constexpr if: [LLVM Bug](https://reviews.llvm.org/D26953).

**Comment by BhaaLseN** [comment]
MSVC doesn't support `if constexpr`, not even with VS 2017 :(

Other than that, why do you need a dedicated NoLpf PB struct? Unless I missed something, the only difference is the `lpf` member (and hence the different padding size); which you could just ignore in the NoLpf case on GC (unless this triggers an unused member warning due to being removed by the `if constexpr`)

**Comment by MerryMage** [comment]
@BhaaLseN The existence of the `lpf` member moves the location/offset of `loop_counter`.

**Comment by BhaaLseN** [comment]
Ah, duh. Forgot about that. Nvm then.

**Comment by JMC47** [comment]
So, with MS not supporting that, what's going to happen to this?

**Comment by MerryMage** [comment]
My plan was to wait until C++17 is a thing in Dolphin and for MSVC `if constexpr` support.

**Comment by BhaaLseN** [comment]
Looking at this again, isn't [this](https://github.com/dolphin-emu/dolphin/blob/master/Source/Core/Core/HW/DSPHLE/UCodes/AXVoice.h#L83) the reason why we have to care about the struct layout (and potentially create duplicates)?
I could imagine an alternate solution where the members are swapped/copied by hand, using the correct offsets for the given AX variant...not sure if this is something we want to consider though.

**Comment by delroth** [comment]
@BhaaLseN that's what I ended up doing for Zelda HLE, see https://github.com/dolphin-emu/dolphin/blob/master/Source/Core/Core/HW/DSPHLE/UCodes/Zelda.cpp#L885

I'm not against doing the same thing for AX.

**Comment by shuffle2** [comment]
can't you just use unions for this (instead of the #ifdef stuff)?

**Comment by MerryMage** [comment]
Updated to use copies for the alternate layout.

File: Source/Core/Core/HW/DSPHLE/UCodes/AX.cpp [hunk 1]
```diff
@@ -22,6 +22,17 @@ namespace DSP
 {
 namespace HLE
 {
+static bool HasLpf(u32 crc)
+{
+  switch (crc)
+  {
+  case 0x4e8a8b21:
```
Comment (Warepire) [inline Source/Core/Core/HW/DSPHLE/UCodes/AX.cpp]: Is this documented somewhere else? It looks highly magic to me.

Comment (JosJuice) [inline Source/Core/Core/HW/DSPHLE/UCodes/AX.cpp]: Yes, the ucode hashes are listed at https://github.com/dolphin-emu/dolphin/blob/7fa14169e9aafd011cbb658e32a259e701812be0/Source/Core/Core/HW/DSPHLE/UCodes/UCodes.cpp#L252

Comment (Warepire) [inline Source/Core/Core/HW/DSPHLE/UCodes/AX.cpp]: Could that file be referenced, or could the check be performed in that file instead (assuming that works with how the code is designed)?

Comment (MerryMage) [inline Source/Core/Core/HW/DSPHLE/UCodes/AX.cpp]: UCode variations should be handed by each individual UCode, imo.

See the [Zelda UCode](https://github.com/dolphin-emu/dolphin/blob/7fa14169e9aafd011cbb658e32a259e701812be0/Source/Core/Core/HW/DSPHLE/UCodes/Zelda.cpp#L75) for an example. I feel this matches the intended usecase for the m_crc member.

Comment (Warepire) [inline Source/Core/Core/HW/DSPHLE/UCodes/AX.cpp]: Alright.

File: Source/Core/Core/HW/DSPHLE/UCodes/AXStructs.h [hunk 2]
```diff
@@ -253,8 +253,9 @@ struct AXPB
   PBSampleRateConverter src;
   PBADPCMLoopInfo adpcm_loop_info;
   PBLowPassFilter lpf;
```
Comment (BhaaLseN) [inline Source/Core/Core/HW/DSPHLE/UCodes/AXStructs.h]: Maybe we should document that `lpf` is a member thats treated differently in certain cases? Least we should do is put a comment similar to "skipped when writing to/reading from MRAM/ARAM for _certain_ AX UCodes"

Comment (MerryMage) [inline Source/Core/Core/HW/DSPHLE/UCodes/AXStructs.h]: Makes sense, done.

**Comment by JMC47** [comment]
LGTM.  I completed the first three missions of RS2 (reported to crash with the hack in Ishiiruka) and they seem to be working fine here.  Rogue Squadron 3 isn't working well in Dolphin right now, but, in the parts of the game I got to it also works.

As a bonus, it appears that Pac-Man World Rally is fixed.  The audio now seems to be roughly the same as LLE.  Instead of having some sounds super loud, and some super quiet, they all now play fairly loud, but can be adjusted in the options.  The game appears louder than most on console as well.

File: Source/Core/Core/HW/DSPHLE/UCodes/AXVoice.h [hunk 3]
```diff
@@ -81,17 +81,50 @@ union AXBuffers
 };
 
 // Read a PB from MRAM/ARAM
-void ReadPB(u32 addr, PB_TYPE& pb)
+void ReadPB(u32 addr, PB_TYPE& pb, bool has_lpf = true)
```
Comment (delroth) [inline Source/Core/Core/HW/DSPHLE/UCodes/AXVoice.h]: Slight cosmetic preference: I would prefer if ReadPB/WritePB took the CRC as input and then computed HasLpf internally.

It also makes little sense to me to have a default value here -- that's just going to hide bugs.

File: Source/Core/Core/HW/DSPHLE/UCodes/AX.cpp [hunk 4]
```diff
@@ -22,6 +22,17 @@ namespace DSP
...
+    return false;
+  default:
```
Comment (delroth) [inline Source/Core/Core/HW/DSPHLE/UCodes/AX.cpp]: I'm not convinced that the RS2 UCode is the only one to not have LPF support. I'll try to have a look through my research notes...

File: Source/Core/Core/HW/DSPHLE/UCodes/AXVoice.h [hunk 5]
```diff
@@ -80,18 +80,63 @@ union AXBuffers
 #endif
 };
 
+// Determines if this version of the UCode has a PBLowPassFilter in its AXPB layout.
+bool HasLpf(u32 crc)
```
Comment (BhaaLseN) [inline Source/Core/Core/HW/DSPHLE/UCodes/AXVoice.h]: Maybe this could/should be static?

Comment (MerryMage) [inline Source/Core/Core/HW/DSPHLE/UCodes/AXVoice.h]: It's in an anonymous namespace; it's already static.

**Comment by delroth** [comment]
I'm fine with merging this now because even if other UCode versions have the same weirdness, it should only impact LPF emulation which is presently disabled due to other issues. I think it's a good compromise.

**Pull request merged**